File: python/src/chiptime/metrics/sports.py
"""Sport profiles — the table that knows how each sport measures itself.

Profiles are data, not subclasses (ADR-0008 §2): analytics code branches on
profile fields, never on sport names scattered through logic. Mapping per
docs/research/sport-metrics-domain.md §0.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from chiptime.model import Session

PaceStyle = Literal["per_km", "per_100m", "per_500m", "speed"]

# Below this, a running cadence is per-leg strides/min on many devices
# (taxonomy #66); display convention is steps/min, so we double and label.
RUN_PER_LEG_CADENCE_MAX = 130.0


@dataclass(frozen=True, slots=True)
class SportProfile:
    """How one sport measures itself — profiles are data, not subclasses.

    Analytics code branches on these fields, never on sport names, so adding
    a sport is a table row, not a code path.

    Attributes:
        key: Profile name (``"running"``, ``"pool_swim"``, ...).
        pace_style: How speed is presented — ``per_km``, ``per_100m``,
            ``per_500m``, or ``speed`` (km/h).
        primary: Preferred intensity signal when its stream exists
            (``"power"`` or ``"speed"``).
        cadence_units: Display convention (``rpm``, ``spm``, strokes/min).
        cadence_double_if_per_leg: Running heuristic — cadence below 130 is
            per-leg strides on many devices; doubled for display, labeled.
        distance_from_lengths: Pool truth — distance is lengths x pool size,
            never GPS.
    """

    key: str
    pace_style: PaceStyle
    primary: Literal["power", "speed"]  # preferred intensity signal when present
    cadence_units: str  # display convention: rpm | spm | strokes/min
    cadence_double_if_per_leg: bool = False
    distance_from_lengths: bool = False  # pool truth: lengths x pool size, never GPS


RUNNING = SportProfile("running", "per_km", "speed", "spm", cadence_double_if_per_leg=True)
CYCLING = SportProfile("cycling", "speed", "power", "rpm")
POOL_SWIM = SportProfile(
    "pool_swim", "per_100m", "speed", "strokes/min", distance_from_lengths=True
)
OPEN_WATER_SWIM = SportProfile("open_water_swim", "per_100m", "speed", "strokes/min")
ROWING = SportProfile("rowing", "per_500m", "power", "strokes/min")
HIKING = SportProfile("hiking", "per_km", "speed", "spm")
XC_SKIING = SportProfile("cross_country_skiing", "per_km", "speed", "spm")
GENERIC = SportProfile("generic", "per_km", "speed", "rpm")

_BY_SPORT: dict[str, SportProfile] = {
    "running": RUNNING,
    "cycling": CYCLING,
    "rowing": ROWING,
    "hiking": HIKING,
    "walking": HIKING,
    "cross_country_skiing": XC_SKIING,
}


def profile_for(session: Session) -> SportProfile:
    """Resolve (sport, sub_sport) → profile; unknown sports get GENERIC
    (correct-but-shallow beats wrong-but-specific)."""
    sport = (session.sport or "").lower()
    sub = (session.sub_sport or "").lower()
    if sport == "swimming":
        return OPEN_WATER_SWIM if sub == "open_water" else POOL_SWIM
    if sub == "indoor_rowing" or (sport == "fitness_equipment" and "row" in sub):
        return ROWING
    return _BY_SPORT.get(sport, GENERIC)


def primary_signal(session: Session) -> tuple[str, str | None]:
    """The intensity signal actually available: profile preference constrained
    by which streams exist. Returns (kind, stream_name); kind is
    "power" | "speed" | "none"."""
    p = profile_for(session)
    if p.primary == "power":
        s = session.records.stream("power")
        if s is not None and s.present_count > 0:
            return "power", "power"
    for name in ("enhanced_speed", "speed"):
        s = session.records.stream(name)
        if s is not None and s.present_count > 0:
            return "speed", name
    return "none", None


def cadence_display(
    avg_cadence: float | None,
    profile: SportProfile,
) -> tuple[float | None, str, str | None]:
    """(value, units, note). The doubling heuristic is labeled, never silent."""
    if avg_cadence is None:
        return None, profile.cadence_units, None
    if profile.cadence_double_if_per_leg and avg_cadence < RUN_PER_LEG_CADENCE_MAX:
        return avg_cadence * 2.0, profile.cadence_units, "doubled_per_leg_cadence"
    return avg_cadence, profile.cadence_units, None

File: js/src/metrics/sports.ts
/**
 * Sport profiles — the table that knows how each sport measures itself.
 *
 * Twin of `python/src/chiptime/metrics/sports.py`. Profiles are data, not
 * subclasses (ADR-0008 §2).
 */

import { type Session, getStream, presentCount } from "../model.js";

export type PaceStyle = "per_km" | "per_100m" | "per_500m" | "speed";

/**
 * Below this, a running cadence is per-leg strides/min on many devices (taxonomy
 * #66); display convention is steps/min, so we double and label.
 */
export const RUN_PER_LEG_CADENCE_MAX = 130.0;

export interface SportProfile {
  readonly key: string;
  readonly paceStyle: PaceStyle;
  readonly primary: "power" | "speed";
  readonly cadenceUnits: string;
  readonly cadenceDoubleIfPerLeg: boolean;
  readonly distanceFromLengths: boolean;
}

function sp(
  key: string,
  paceStyle: PaceStyle,
  primary: "power" | "speed",
  cadenceUnits: string,
  opts: { double?: boolean; lengths?: boolean } = {},
): SportProfile {
  return {
    key,
    paceStyle,
    primary,
    cadenceUnits,
    cadenceDoubleIfPerLeg: opts.double ?? false,
    distanceFromLengths: opts.lengths ?? false,
  };
}

export const RUNNING = sp("running", "per_km", "speed", "spm", { double: true });
export const CYCLING = sp("cycling", "speed", "power", "rpm");
export const POOL_SWIM = sp("pool_swim", "per_100m", "speed", "strokes/min", { lengths: true });
export const OPEN_WATER_SWIM = sp("open_water_swim", "per_100m", "speed", "strokes/min");
export const ROWING = sp("rowing", "per_500m", "power", "strokes/min");
export const HIKING = sp("hiking", "per_km", "speed", "spm");
export const XC_SKIING = sp("cross_country_skiing", "per_km", "speed", "spm");
export const GENERIC = sp("generic", "per_km", "speed", "rpm");

const BY_SPORT: Readonly<Record<string, SportProfile>> = {
  running: RUNNING,
  cycling: CYCLING,
  rowing: ROWING,
  hiking: HIKING,
  walking: HIKING,
  cross_country_skiing: XC_SKIING,
};

/**
 * Resolve (sport, subSport) → profile; unknown sports get GENERIC
 * (correct-but-shallow beats wrong-but-specific).
 */
export function profileFor(session: Session): SportProfile {
  const sport = (session.sport ?? "").toLowerCase();
  const sub = (session.subSport ?? "").toLowerCase();
  if (sport === "swimming") return sub === "open_water" ? OPEN_WATER_SWIM : POOL_SWIM;
  if (sub === "indoor_rowing" || (sport === "fitness_equipment" && sub.includes("row"))) {
    return ROWING;
  }
  return BY_SPORT[sport] ?? GENERIC;
}

/**
 * The intensity signal actually available: profile preference constrained by which
 * streams exist. Returns [kind, streamName]; kind is "power" | "speed" | "none".
 */
export function primarySignal(session: Session): [string, string | null] {
  const p = profileFor(session);
  if (p.primary === "power") {
    const s = getStream(session.records, "power");
    if (s !== undefined && presentCount(s) > 0) return ["power", "power"];
  }
  for (const name of ["enhanced_speed", "speed"]) {
    const s = getStream(session.records, name);
    if (s !== undefined && presentCount(s) > 0) return ["speed", name];
  }
  return ["none", null];
}

/** [value, units, note]. The doubling heuristic is labeled, never silent. */
export function cadenceDisplay(
  avgCadence: number | null,
  profile: SportProfile,
): [number | null, string, string | null] {
  if (avgCadence === null) return [null, profile.cadenceUnits, null];
  if (profile.cadenceDoubleIfPerLeg && avgCadence < RUN_PER_LEG_CADENCE_MAX) {
    return [avgCadence * 2.0, profile.cadenceUnits, "doubled_per_leg_cadence"];
  }
  return [avgCadence, profile.cadenceUnits, null];
}
